### ISSUE TITLE
Evolved anchors return the same default anchors size

### ISSUE THREAD

**Issue by kimsung-git** [ultralytics/yolov5#3051]
## ❔Question
As far as I understood, autoanchor evolves and finds the most optimal anchors size from customized dataset so we don't have to modify the anchor size value by hand. However, after training from scratch with customized dataset, anchors size seem the same as what is defined in yolov5l.yaml. 

This is the command I used for training
```python
nohup python -m torch.distributed.launch --nproc_per_node 4 \
    train.py \
    --batch-size 64 \
    --img 640 \
    --epochs 500 \
    --data airport.yaml \
    --project airport \
    --weights '' \
    --cfg models/yolov5l.yaml  & 
```

Also this is command I used to check anchors size
```python
model = torch.load(opt.weight, map_location=None)  # load
    m = model['model'].model[-1]  # Detect() layer
    print(m.anchor_grid.squeeze())  # print anchors
```

And this is the size of anchors
```python
tensor([[[ 10.,  13.],
         [ 16.,  30.],
         [ 33.,  23.]],

        [[ 30.,  61.],
         [ 62.,  45.],
         [ 59., 119.]],

        [[116.,  90.],
         [156., 198.],
         [373., 326.]]], device='cuda:0', dtype=torch.float16)
```

The size of anchor didn't change but same as the default values which is from COCO dataset. Does this mean autoanchor didn't work properly?

## Additional context
Thanks for your contribution to this work.

**Comment by kimsung-git** [comment]
After reading over hyperparameter tuning tutorial, it seems that I need to uncomment anchors parameter at hyp.scratch.yaml for autoanchor to overwrite anchor information in yolov5l.yaml(actually to activate autoanchor properly). So the reason why I had the same anchors size was because the program just read anchor information from yolov5l.yaml and didn't really run autoanchor.

I get different autoanchor log with or without anchors parameter at hyp.scratch.yaml 

with anchors parameter 
```linux
autoanchor: Analyzing anchors... anchors/target = 0.32, Best Possible Recall (BPR) = 0.0839. Attempting to improve anchors, please wait...
autoanchor: WARNING: Extremely small objects found. 1302 of 490147 labels are < 3 pixels in size.
autoanchor: Running kmeans for 9 anchors on 490138 points...
autoanchor: thr=0.25: 0.9866 best possible recall, 4.08 anchors past thr
autoanchor: n=9, img_size=640, metric_all=0.279/0.665-mean/best, past_thr=0.473-mean: 17,22,  44,54,  55,120,  139,93,  102,213,  159,345,  290,208,  296,491,  527,350
autoanchor: Evolving anchors with Genetic Algorithm: fitness = 0.7008: 100%|██████████| 1000/1000 [00:32<00:00, 30.69it/s]
autoanchor: thr=0.25: 0.9976 best possible recall, 4.34 anchors past thr
autoanchor: n=9, img_size=640, metric_all=0.292/0.701-mean/best, past_thr=0.473-mean: 10,15,  25,24,  22,55,  60,46,  50,118,  123,111,  92,228,  184,314,  422,370
autoanchor: New anchors saved to model. Update model *.yaml to use these anchors in the future.
```

without anchors parameter
```linux 
autoanchor: Analyzing anchors... anchors/target = 4.44, Best Possible Recall (BPR) = 0.9980
```

So the solution to my question is to make sure autoanchor run by uncommenting anchors parameter in hyp.scratch.yaml?

**Comment by glenn-jocher** [comment]
@kimsung-git autoanchor analyzes your anchors in the context of your dataset and training settings, and will evolve you new anchors only if your current anchors are missing or they are determined to be a poor fit based on the BPR metric.

**Comment by glenn-jocher** [comment]
@kimsung-git also if autoanchor completes with no error then it is working correctly.

**Comment by kimsung-git** [comment]
@glenn-jocher what do you mean by “if you’re current  anchors are missing?” You mean when anchors parameter is not set in hyp.scratch.yaml? Thanks for the help!

**Comment by glenn-jocher** [comment]
@kimsung-git anchor definition is optional. Model yaml files support either explicit anchor definition or specifying an anchor count per layer. If your hyp file defines an anchor count it will take precedence over your model yaml anchors.
```yaml
anchors: 3
```
